File: Array/210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        

        Map<Integer, List<Integer>> graph = new HashMap<>();
        Map<Integer, Integer> indegree = new HashMap<>();

        for(int i=0;i<numCourses;i++) {
            indegree.put(i, 0);
        }


        for(int[] prereq: prerequisites) {
            int v = prereq[0];
            int u = prereq[1];

            graph.putIfAbsent(u, new ArrayList<>());
            List<Integer> nodes = graph.get(u);

            nodes.add(v);
            indegree.put(v, indegree.get(v)+1);
        }

        Queue<Integer>q = new LinkedList<>();

        for(int key: indegree.keySet()) {
            if(indegree.get(key) == 0) {
                q.offer(key);
            }
        }



        List<Integer> out = new ArrayList<>();

        while(!q.isEmpty()) {
            int node = q.poll();

            out.add(node);

            List<Integer> neighbours = graph.getOrDefault(node, new ArrayList<>());

            for(int nei: neighbours) {
                indegree.put(nei, indegree.get(nei)-1);
                if(indegree.get(nei) == 0) {
                    q.offer(nei);
                }
            }
            
        }


        return (out.size() != numCourses) ? new int[]{} 
            : out.stream().mapToInt(Integer::intValue).toArray();

    }
}

File: Array/210-course-schedule-ii/course-schedule-ii.py
from collections import deque
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegree = {i:0 for i in range(numCourses)}
        graph = {i: [] for i in range(numCourses)}

        for v, u in prerequisites:
            graph[u].append(v)
            indegree[v]+=1

        q = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
        

        result = []
        while q:
            node = q.popleft()
            result.append(node)

            for nei in graph[node]:
                indegree[nei]-=1
                if indegree[nei] == 0:
                    q.append(nei)

        return result if len(result) == numCourses else []